### ISSUE TITLE
EdDSA fail to sign

### ISSUE THREAD

**Issue by HDNAdministrator** [web-token/jwt-framework#356]
**Version(s) affected**: 3.0.6

**Description**
JWS signature fails when using EdDSA

**How to reproduce**

`

		$jwk= JWKFactory::createOKPKey('Ed25519');

		$payload = json_encode([
			'iat' => time(),
			'nbf' => time(),
			'exp' => time() + 3600,
			'iss' => 'ISS',
			'aud' => 'AUD'
		]);

		$algorithmManager = new AlgorithmManager([new EdDSA()]);

		$jwsBuilder = new JWSBuilder($algorithmManager);

		$jws = $jwsBuilder
				->create()
				->withPaylod($payload)
				->addSignature($jwk)
				->build();
`

**Possible Solution**
[sodium-crypto-sign-detached](https://www.php.net/manual/en/function.sodium-crypto-sign-detached.php) on line 48 in EdDSA use the secret key 'd' and not the concatenation of the secret and public key 'd' and 'x'.
Lines from 35 to 38, 43 and 45 aren't necessary as consequence.

**Additional context**

**Comment by Spomky** [comment]
😬 I should verify the test vectors at the same time.

**Comment by HDNAdministrator** [comment]
I tested sucessfully.

**Comment by Spomky** [comment]
I cannot reproduce this issue.
The secret shall be the concatenation of the public and private keys otherwise the following exception is thrown:

> PHP Fatal error:  Uncaught SodiumException: sodium_crypto_sign_detached(): Argument #2 ($secret_key) must be SODIUM_CRYPTO_SIGN_SECRETKEYBYTES bytes long

`SODIUM_CRYPTO_SIGN_SECRETKEYBYTES` corresponds to 64 bytes long 

This following code works as expected (computation and verification):

```php

use Jose\Component\Core\AlgorithmManager;
use Jose\Component\KeyManagement\JWKFactory;
use Jose\Component\Signature\Algorithm\EdDSA;
use Jose\Component\Signature\JWSBuilder;
use Jose\Component\Signature\JWSLoader;
use Jose\Component\Signature\JWSVerifier;
use Jose\Component\Signature\Serializer\CompactSerializer;
use Jose\Component\Signature\Serializer\JWSSerializerManager;

$algorithmManager = new AlgorithmManager([new EdDSA()]);
$jwsBuilder = new JWSBuilder($algorithmManager);
$serializer = new CompactSerializer();
$serializerManager = new JWSSerializerManager([$serializer]);
$jwsVerifier = new JWSVerifier($algorithmManager);

$jwk= JWKFactory::createOKPKey('Ed25519');
$payload = json_encode([
    'iat' => time(),
    'nbf' => time(),
    'exp' => time() + 3600,
    'iss' => 'ISS',
    'aud' => 'AUD'
]);

$jws = $jwsBuilder
    ->create()
    ->withPayload($payload)
    ->addSignature($jwk, ['alg' => 'EdDSA'])
    ->build();

$token = $serializer->serialize($jws);

$jwsLoader = new JWSLoader($serializerManager, $jwsVerifier, null);

$signature = null;
dd($jwsLoader->loadAndVerifyWithKey($token, $jwk, $signature)->getPayload());
```

**Comment by HDNAdministrator** [comment]
That is strange. I've reran the test code with the original and the suggestion versions and I still get the same exception that you have but on the original. By [php doc](https://www.php.net/manual/en/function.sodium-crypto-sign.php) it is suppose to be the privateKey only not the concat. Could it be related with the version of sodium?

From phpinfo:
- PHP: 8.1.8
- Sodium: 1.0.18

EDIT: I apologise, I forgot to mention that JWKFactory was modified too. Probably the source of the difference.
The racional is, given [sodium gives us the private key](https://www.php.net/manual/en/function.sodium-crypto-sign-secretkey.php), aren't the next couple lines of code unnecessary (rigth after the switch statement)?

Below the JWKFactory.createOKPKey modified version.

`

    public static function createOKPKey(string $curve, array $values = []): JWK
    {
        if (! extension_loaded('sodium')) {
            throw new RuntimeException('The extension "sodium" is not available. Please install it to use this method');
        }

        switch ($curve) {
            case 'X25519':
                $keyPair = sodium_crypto_box_keypair();
                // $secret = sodium_crypto_box_secretkey($keyPair);
                $d = sodium_crypto_box_secretkey($keyPair);
                $x = sodium_crypto_box_publickey($keyPair);

                break;

            case 'Ed25519':
                $keyPair = sodium_crypto_sign_keypair();
                // $secret = sodium_crypto_sign_secretkey($keyPair);
                $d = sodium_crypto_sign_secretkey($keyPair);
                $x = sodium_crypto_sign_publickey($keyPair);

                break;

            default:
                throw new InvalidArgumentException(sprintf('Unsupported "%s" curve', $curve));
        }
        // $secretLength = mb_strlen($secret, '8bit');
        // $d = mb_substr($secret, 0, -$secretLength / 2, '8bit');

        $values = array_merge(
            $values,
            [
                'kty' => 'OKP',
                'crv' => $curve,
                'd' => Base64UrlSafe::encodeUnpadded($d),
                'x' => Base64UrlSafe::encodeUnpadded($x),
            ]
        );

        return new JWK($values);
    }

`

**Comment by HDNAdministrator** [comment]
TL;DR: Ed25519 and X25519 are generated differently on ECDHES and JWKFactory

On ECDHES
`

    private function createOKPKey(string $curve): JWK
    {
        $this->checkSodiumExtensionIsAvailable();

        switch ($curve) {
            case 'X25519':
                $keyPair = sodium_crypto_box_keypair();
                $d = sodium_crypto_box_secretkey($keyPair);
                $x = sodium_crypto_box_publickey($keyPair);

                break;

            case 'Ed25519':
                $keyPair = sodium_crypto_sign_keypair();
                $d = sodium_crypto_sign_secretkey($keyPair);
                $x = sodium_crypto_sign_publickey($keyPair);

                break;

            default:
                throw new InvalidArgumentException(sprintf('Unsupported "%s" curve', $curve));
        }

        return new JWK([
            'kty' => 'OKP',
            'crv' => $curve,
            'x' => Base64UrlSafe::encodeUnpadded($x),
            'd' => Base64UrlSafe::encodeUnpadded($d),
        ]);
    }
`
On JWKFactory
`

    public static function createOKPKey(string $curve, array $values = []): JWK
    {
        if (! extension_loaded('sodium')) {
            throw new RuntimeException('The extension "sodium" is not available. Please install it to use this method');
        }

        switch ($curve) {
            case 'X25519':
                $keyPair = sodium_crypto_box_keypair();
                $secret = sodium_crypto_box_secretkey($keyPair);
                $x = sodium_crypto_box_publickey($keyPair);

                break;

            case 'Ed25519':
                $keyPair = sodium_crypto_sign_keypair();
                $secret = sodium_crypto_sign_secretkey($keyPair);
                $x = sodium_crypto_sign_publickey($keyPair);

                break;

            default:
                throw new InvalidArgumentException(sprintf('Unsupported "%s" curve', $curve));
        }
        $secretLength = mb_strlen($secret, '8bit');
        $d = mb_substr($secret, 0, -$secretLength / 2, '8bit');

        $values = array_merge(
            $values,
            [
                'kty' => 'OKP',
                'crv' => $curve,
                'd' => Base64UrlSafe::encodeUnpadded($d),
                'x' => Base64UrlSafe::encodeUnpadded($x),
            ]
        );

        return new JWK($values);
    }
`

**Comment by Spomky** [comment]
Correct. I will fix this issue ASAP

**Comment by Spomky** [comment]
Hi,

This should be fixed in 3.1.1.

**Comment by Spomky** [comment]
By the way, many thanks for supporting me 👍🏽 . I really² appreciate 😍

**Comment by HDNAdministrator** [comment]
The lib is great to the point I’m using it as guidance with the offical docs to build one in kotlin Enviado do Correio para Windows De: Florent MorselliEnviado: 24 de agosto de 2022 07:50Para: web-token/jwt-frameworkCc: HDN; AuthorAssunto: Re: [web-token/jwt-framework] EdDSA fail to sign (Issue #356) By the way, many thanks for supporting me 👍🏽 . I really² appreciate 😍—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you authored the thread.Message ID: ***@***.***>
